dlang/source/flatbuffers/flatbufferbuilder.d

```diff
@@ -0,0 +1,301 @@
+module flatbuffers.flatbufferbuilder;
+
+import flatbuffers.exception;
+import flatbuffers.bytebuffer;
+
+import std.exception;
+
+final class FlatBufferBuilder
+{
+public: 
+	this(int initsize)
+	{
+		if(initsize == 0)
+			throw new ArgumentOutOfRangeException("initsize", initsize, "Must be greater than zero");
+		_space = initsize;
+		_buffer = new ByteBuffer(new ubyte[initsize]);
+	}
+	
+	int offset() { return _buffer.length - _space; }
+	
+	void pad(int size)
+	{
+		for(int i=0; i<size; i++) {
+			-- _space;
+			_buffer.put!ubyte(_space, 0x00);
+		}
+	}
+	
+	///Doubles the size of the ByteBuffer, and copies the old data towards
+	///the end of the new buffer (since we build the buffer backwards).
+	void growBuffer()
+	{
+		auto oldBuf = _buffer.data;
+		auto oldBufSize = oldBuf.length;
+		if((oldBufSize & 0xC0000000) != 0)
+			throw new Exception("FlatBuffers: cannot grow buffer beyond 2 gigabytes.");
+		
+		auto newBufSize = oldBufSize * 2;
+		auto newBuf = new ubyte[](newBufSize);
+		newBuf[(newBufSize-oldBufSize)..$] = oldBuf[];
+		
+		_buffer = new ByteBuffer(newBuf);
+	}
+	
+	///Prepare to write an element of `size` after `additional_bytes`
+	///have been written, e.g. if you write a string, you need to align
+	///such the int length field is aligned to SIZEOF_INT, and the string
+	///data follows it directly.
+	///If all you need to do is align, `additional_bytes` will be 0.
+	void prep(int size, int additionalBytes)
+	{
+		//Track the biggest thing we've ever aligned to.
+		if(size > _minAlign)
+			_minAlign = size;
+		
+		//Find the amount of alignment needed such that `size` is properly
+		//aligned after `additional_bytes`.
+		auto alignSize = ((~(cast(int)_buffer.length - _space + additionalBytes)) + 1) & (size - 1);
+		
+		//Reallocate the buffer if needed.
+		while(_space < alignSize + size + additionalBytes)
+		{
+			auto oldBufSize = cast(int)_buffer.length;
+			growBuffer();
+			_space += cast(int)_buffer.length - oldBufSize;
+		}
+		if(alignSize > 0)
+			pad(alignSize);
+	}
+
+	void put(T)(T x) if(is(T == bool) || is(T == byte) || is(T == ubyte) ||
+		is(T == short) || is(T == ushort) || is(T == int) || is(T == uint) || is(T == long) 
+		|| is(T == ulong) || is(T == float) || is(T == double)) 
```

Comment:
any reason for not using `std.traits.isNumeric!T || is(T == bool)` instead of all these `is(T==type)` checks?